### ISSUE TITLE
"Sequence contains no matching element" in RealmAdapter.Get(long)

### ISSUE THREAD

**Issue by amulware** [toggl/mobileapp#2920]
## 💥 Crash Report
It crashes when fetching some object by Id. We're not doing that in a lot of places, are we? Maybe this alone is enough info to deduce the problem.

### 👥 Impact
- 21 times to 19 users in 1.8.1 so far, that's about 15% of crashes
- different phones and OS versions

### 🔗 Context
I think we were trying to give more meaningful exceptions to some of these cases, to for example resolve the generic type, so we know what is being fetched here. Maybe we have to do that here too, if we don't know how else to fix it.

### 🛰 Link
https://appcenter.ms/orgs/Toggl/apps/Toggl-iOS/crashes/groups/a83cce09bc69a9d270bb98010728a8ef3e168475?appVersion=1.8.1

### 📝  Exception & Stack Trace
`System.InvalidOperationException: Sequence contains no matching element`
```cs
RealmResultsVisitor.VisitMethodCall (System.Linq.Expressions.MethodCallExpression node)
MethodCallExpression.Accept (System.Linq.Expressions.ExpressionVisitor visitor)
ExpressionVisitor.Visit (System.Linq.Expressions.Expression node)
RealmResultsProvider.Execute (System.Linq.Expressions.Expression expression)
RealmResultsProvider.Execute[T] (System.Linq.Expressions.Expression expression)
Queryable.Single[TSource] (System.Linq.IQueryable`1[T] source, System.Linq.Expressions.Expression`1[TDelegate] predicate)
RealmAdapter`2[TRealmEntity,TModel].Get (System.Int64 id)
Repository`1+<>c__DisplayClass1_0[TModel].<GetById>b__0 ()
BaseStorage`1+<>c__DisplayClass10_0`1[TModel,T].<CreateObservable>b__0 (System.IObserver`1[T] observer)
```

**Comment by khoi** [comment]
Unassigning myself cause after looking into it, I don't really know what to do 😅

**Comment by sofokli1** [comment]
my theory for this issue is related to what the new PR #3176 is trying to solve. Although i can't prove it, logic dictates that the issue is related to some kind of race condition. Statistically speaking related to TEs more than anything else. Looks like a race condition where the id has changed due to sync or deletion or other. I would suggest to see how it will behave with PR #3176 before trying other stuff

**Comment by amulware** [comment]
I agree with the above, let's see if this issue persists in 1.10 :)

**Comment by dshr** [comment]
This also affects Android builds: https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/groups/250e4d39dc1bd7518eb93c911db65cd31fc09f58?appVersion=1.3.2&period=last90Days

**Comment by amulware** [comment]
Still a thing in 1.13.1, and currently by far the most common symbolicated crash.

https://appcenter.ms/orgs/Toggl/apps/Toggl-iOS/crashes/groups/e898b463e23316015d1b31958d2209192dcb85b0?appVersion=1.13.1

If we don't know how to fix this right now, then I want the following:
- find all places in the code that can possibly cause this crash
	- **list them in this issue**
- catch the exception in all those places
- if it's in a place where it can happen by design, handle the error gracefully if easily possible (and if it doesn't cause any glitches or unexpected behaviour), and log the error _with additional information_
- **otherwise** rethrow it _with additional information_

Additional info I can think of right now:
- identify which of the places it is, either by some unique string ID, or just describing the action that is happening
- the actual entity type it's trying to fetch (`TModel` isn't very helpful)

**Comment by amulware** [comment]
Piling these on here in case they are different cases (one of the stack traces mentions the edit view):
https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/errors/141257325u/overview
https://appcenter.ms/orgs/Toggl/apps/Toggl-Android/crashes/errors/778203783u/overview

**Comment by simonrozsival** [comment]
All places where we call `Repository.GetById`:
- `ReportsProvider` on line `149` - all exceptions should be caught on line `152` - to verify this, we should test that reports show correctly archived projects
- `IsBillableAvailableForProjectInteractor`
- `ProjectDefaultsToBillableInteractor`
- `WorkspaceHasFeatureInteractor`
- `DeleteTimeEntryInteractor`
    - this is my suspect, if there is a time entry which actually doesn't exist shown in the UI and the user tries to delete it, the app will crash
        - this used to happen sometimes in the past on iOS but I am quite confident it doesn't happen anymore but from what @alari76 said earlier, it might still be a problem on Android
        - it should be easy to gracefully handle (and log!) this case
- `EditTimeEntryViewModel` on line `282` - the same problem as ☝️ the previous one. If there is a time entry which does not exist in the log (there is no TE with the ID that is stored in the row) and user taps on the row, opening the edit view would try to get the TE and the app would 💥 
- `GetDefaultWorkspaceInteractor` - invalid op is caught explicitly in the observable query
- `AreProjectsBillableByDefaultInteractor`
- `GetWorkspaceByIdInteractor`
- `UpdateTimeEntryInteractor`
    - is called from `CalendarViewModel` and `EditTimeEntryViewModel`
    - this could crash if the edited time entry or the time entry displayed in the calendar was deleted between initialization of the view and the update action
- `ReportsViewModel` on line `200`
- `StartTimeEntryViewModel` on lines `335` and `347` - starting a time entry based on Siri Shortcut, copying project and tags
    - this could easily be a problem if a shortcut refers to tags which were deleted since the shortcut was deleted
    - this should be easy to handle (ignore tags which don't exist), but we don't handle it at all currently
- `StartTimeEntryViewModel` on line `497` - this should not be a problem, because it tries to fetch a project which was just created using an ID of that newly created project. There is no way that the project could be deleted before we would try to access it again (even if app was paused/terminated between those instructions)
- `EditProjectViewModel` on line `236`, `527`, `534`
    - this could be a potential problem, but it seems highly improbable/a very narrow edge case
    - user would have to start sync, start editing a time entry, go to the client selection view where the list of clients is loaded into the memory when the view opens, then the sync algorithm would delete some clients from our app because they were deleted on server, user would select one of these clients and confirmed it, 💥 
   - the same scenario applies to projects and tasks
- `SelectProjectViewModel` on line `223` - same as with `StartTiemEntryViewModel` line `497`, should not be a problem
- `CalendarViewModel` on line `270` - when a TE is updated by dragging
    - this might also crash if the time entry is deleted by sync while the calendar view is opened (calendar does not react to time entries being created, updated or deleted in the background)

**Comment by simonrozsival** [comment]
There are quite a few places where we call the method. I think that the best way how to fix this situation would be:
- move all usage of data sources into interactors
- require interactors to catch `InvalidOperationException` and wrap it in an exception which gives us more information about where the exception originated
- make data sources internal so that we don't expose them to view models and they can't be used directly without interactors (@heytherewill might like this idea)

I think we should investigate the reports of duplicate entries in the Android app. Maybe there is a bug in `DiffUtil`?

**Comment by Odrakir** [comment]
Ok, my idea is to follow @simonrozsival suggestions one datasource at a time. And as this is a pretty big task I'll be making different PRs for each one of them. How does it sound?

**Comment by amulware** [comment]
This is the direction we're moving in anyways, and each interactor is not a lot of work by itself, right? But we essentially need one pull request for each of the bullet points above?

If @heytherewill agrees this is worth doing right now, then I'd say turn the above list into [ ] checkable items and do as you propose, starting with those cases that seem the most likely to be problematic.

When you write/update the interactors, make sure to test this specific exception, and think about what could go wrong and how we should handle errors. (Fewer crashes are important, but maybe we rather sometimes crash than getting the app into an unusable state, or even worse, corrupt data.)

**Comment by Odrakir** [comment]
I was thinking on one PR per app entity (Workspace, Client, Project...) but I could do it in smaller PRs too.

**Comment by heytherewill** [comment]
I don't think this alone will be enough to make the DataSources internal (mainly because of the observable data sources), I'm afraid, but I really think this is worth doing and will help us in many ways, so let's do it using the checklist approach. 

This is a step in the right direction, for the PrimeRadiant refactor will be easier if the DataSources are contained in interactors.

@Odrakir I think one PR per entity should suffice

**Comment by Odrakir** [comment]
I have a question for @simonrozsival about "require interactors to catch InvalidOperationException and wrap it in an exception which gives us more information about where the exception originated".

What do you mean by "require"? There's no way to make all interactors catch that, right?

**Comment by amulware** [comment]
Removed from projects. Instead, we will refactor interactors one at a time, starting with those that are most likely to cause crashes.

**Comment by Odrakir** [comment]
#3856 lays the base for this to be closed, but now it needs implementation in all those places, so I'm reopening this.

**Comment by amulware** [comment]
This is still happening in 1.7:
https://appcenter.ms/orgs/Toggl/apps/Toggl-iOS/crashes/errors/2146858169u/overview

Can someone confirm that with the changes in #4136 (they are in 1.7, right?) we should now get a different exception if the crash was in any of the `GetById` calls?

If that is so, the crash must be somewhere else. (But I don't see how that's possible with the `RealmAdapter`2[TRealmEntity,TModel].Get` line in the stack trace?)

If that is not so, #4136 didn't serve its purpose for this, and I'd like an explanation of why it didn't (the requirement was clearly mentioned in the PR and linked issues) and a proposal for how to move forward.

**Comment by amulware** [comment]
(Still happening on 1.18)

**Comment by simonrozsival** [comment]
At this point we call the `Get` method only in 2 places:
- the `GetByIdInteractor`
- and `GetProjectSummaryInteractor`

From the events it doesn't seem like the crash happens on the reports page, so it's probably the `GetByIdInteractor`. I'll try to add some more tracking (based on #3856) so that we have more data in a few weeks.

**Comment by simonrozsival** [comment]
A PR with extended tracking was merged into `develop` and it will be part of the next release. I'll unassign myself from this issue now and move it to the next release. We need to collect some data before we can proceed with this issue.

**Comment by simonrozsival** [comment]
So far we tracked this with `1.20`: https://appcenter.ms/orgs/Toggl/apps/Toggl-iOS/crashes/errors/2316932820/reports/2518478325356289999-f5cca390-06e1-4cc4-bdef-5fdb19caf112/stacktrace

There are reports of caught errors in:
- getting workspace by ID
- getting time entry by ID

It's always "sequence contains no matching element". The entity with the given ID is not in the database.

**Comment by amulware** [comment]
How come there are no stack traces? That might help a lot.

**Comment by arturs105** [comment]
This crash (or a very similar one) is still happening on 1.21 and is the crash with the most occurrences. 27 crashes to 5 users. What is interesting, is that the crash so far has only happened on iPhone SE and iPhone 6, both running iOS 12.

There are also 2 stack traces:
```cs
RealmResultsVisitor.VisitMethodCall (System.Linq.Expressions.MethodCallExpression node)
MethodCallExpression.Accept (System.Linq.Expressions.ExpressionVisitor visitor)
ExpressionVisitor.Visit (System.Linq.Expressions.Expression node)
RealmResultsProvider.Execute (System.Linq.Expressions.Expression expression)
RealmResultsProvider.Execute[T] (System.Linq.Expressions.Expression expression)
Queryable.Single[TSource] (System.Linq.IQueryable`1[T] source, System.Linq.Expressions.Expression`1[TDelegate] predicate)
RealmAdapter`2[TRealmEntity,TModel].Get (System.Int64 id)
Repository`1+<>c__DisplayClass1_0[TModel].<GetById>b__0 ()
BaseStorage`1+<>c__DisplayClass10_0`1[TModel,T].<CreateObservable>b__0 (System.IObserver`1[T] observer)
```
```cs
ExceptionServicesImpl.Rethrow (System.Exception exception)
ExceptionHelpers.ThrowIfNotNull (System.Exception exception)
AsyncSubject`1[T].GetResult ()
ReportsViewModel.Initialize ()
MainTabBarViewModel.Initialize ()
MvxNavigationService.Navigate (MvvmCross.ViewModels.MvxViewModelRequest request, MvvmCross.ViewModels.IMvxViewModel viewModel, MvvmCross.ViewModels.IMvxBundle presentationBundle, System.Threading.CancellationToken cancellationToken)
MvxNavigationService.Navigate (System.Type viewModelType, MvvmCross.ViewModels.IMvxBundle presentationBundle)
AppStart`1[TFirstViewModelWhenNotLoggedIn].NavigateToFirstViewModel (System.Object hint)
AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_0 (System.Object state)
NSAsyncSynchronizationContextDispatcher.Apply ()
(wrapper managed-to-native) UIKit.UIApplication.UIApplicationMain(int,string[],intptr,intptr)
UIApplication.Main (System.String[] args, System.IntPtr principal, System.IntPtr delegate)
UIApplication.Main (System.String[] args, System.String principalClassName, System.String delegateClassName)
Application.Main (System.String[] args)
```

In the second stack trace there is `ReportsViewModel.Initialize `, so it might be related to the reports view.

Link to AppCenter: https://appcenter.ms/orgs/Toggl/apps/Toggl-iOS/crashes/errors/3534663872u/overview

**Comment by simonrozsival** [comment]
For 1.21 we tracked this: https://appcenter.ms/orgs/Toggl/apps/Toggl-iOS/crashes/errors/3534663872/overview?q=RealmResultsVisitor&searchType=handlederror

Apparently AppCenter does not attach stack traces to handled errors.

**Comment by amulware** [comment]
There is a box for the stacktrace though - is it just that the exeption we give it simply doesn't have a stacktrace? Or we're using the API wrong somehow?

**Comment by simonrozsival** [comment]
This is what I got today just after login:
```cs
Toggl.Storage.Exceptions.DatabaseOperationException`1[[Toggl.Storage.Models.IDatabaseWorkspace, Toggl.Storage, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]]: Database operation failed [T]: Sequence contains no matching element ---> System.InvalidOperationException: Sequence contains no matching element
  at Realms.RealmResultsVisitor.VisitMethodCall (System.Linq.Expressions.MethodCallExpression node) [0x003d8] in <53db4e918d8341cdae1248b5d4c1c726>:0
  at System.Linq.Expressions.MethodCallExpression.Accept (System.Linq.Expressions.ExpressionVisitor visitor) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/external/corefx/src/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs:109
  at System.Linq.Expressions.ExpressionVisitor.Visit (System.Linq.Expressions.Expression node) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/external/corefx/src/System.Linq.Expressions/src/System/Linq/Expressions/ExpressionVisitor.cs:34
  at Realms.RealmResultsProvider.Execute (System.Linq.Expressions.Expression expression) [0x00013] in <53db4e918d8341cdae1248b5d4c1c726>:0
  at Realms.RealmResultsProvider.Execute[T] (System.Linq.Expressions.Expression expression) [0x00000] in <53db4e918d8341cdae1248b5d4c1c726>:0
  at System.Linq.Queryable.Single[TSource] (System.Linq.IQueryable`1[T] source, System.Linq.Expressions.Expression`1[TDelegate] predicate) [0x0001c] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/external/corefx/src/System.Linq.Queryable/src/System/Linq/Queryable.cs:813
  at Toggl.Storage.Realm.RealmAdapter`2[TRealmEntity,TModel].Get (System.Int64 id) [0x00000] in /Users/simonrozsival/Projects/github/toggl/mobileapp/Toggl.Storage.Realm/RealmAdapter.cs:73
  at Toggl.Storage.Realm.Repository`1+<>c__DisplayClass1_0[TModel].<GetById>b__0 () [0x00000] in /Users/simonrozsival/Projects/github/toggl/mobileapp/Toggl.Storage.Realm/Repository.cs:17
  at Toggl.Storage.Realm.BaseStorage`1+<>c__DisplayClass10_0`1[TModel,T].<CreateObservable>b__0 (System.IObserver`1[T] observer) [0x00002] in /Users/simonrozsival/Projects/github/toggl/mobileapp/Toggl.Storage.Realm/BaseStorage.cs:71
  --- End of inner exception stack trace ---
  at System.Reactive.PlatformServices.ExceptionServicesImpl.Rethrow (System.Exception exception) [0x00006] in <12cae72f24674302a4c964a96bed8f31>:0
  at System.Reactive.ExceptionHelpers.ThrowIfNotNull (System.Exception exception) [0x0000d] in <12cae72f24674302a4c964a96bed8f31>:0
  at System.Reactive.Subjects.AsyncSubject`1[T].GetResult () [0x00043] in <12cae72f24674302a4c964a96bed8f31>:0
  at Toggl.Core.UI.ViewModels.Reports.ReportsViewModel.Initialize () [0x0011d] in /Users/simonrozsival/Projects/github/toggl/mobileapp/Toggl.Core.UI/ViewModels/Reports/ReportsViewModel.cs:204
  at Toggl.Core.UI.ViewModels.MainTabBarViewModel.Initialize () [0x000c5] in /Users/simonrozsival/Projects/github/toggl/mobileapp/Toggl.Core.UI/ViewModels/MainTabBarViewModel.cs:163
  at MvvmCross.Navigation.MvxNavigationService.Navigate (MvvmCross.ViewModels.MvxViewModelRequest request, MvvmCross.ViewModels.IMvxViewModel viewModel, MvvmCross.ViewModels.IMvxBundle presentationBundle, System.Threading.CancellationToken cancellationToken) [0x000c1] in <a6e263315c3e4a93a1b6bbd88bb441a4>:0
  at MvvmCross.Navigation.MvxNavigationService.Navigate (System.Type viewModelType, MvvmCross.ViewModels.IMvxBundle presentationBundle) [0x000bd] in <a6e263315c3e4a93a1b6bbd88bb441a4>:0
  at Toggl.Core.UI.ViewModels.LoginViewModel.onAuthenticated () [0x000cb] in /Users/simonrozsival/Projects/github/toggl/mobileapp/Toggl.Core.UI/ViewModels/LoginViewModel.cs:256
  at System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__6_1 (System.Object state) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/runtime/compilerservices/AsyncMethodBuilder.cs:1039
  at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) [0x0000d] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1308
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00071] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:961
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) [0x00000] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/executioncontext.cs:908
  at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00021] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1285
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:858
  at ObjCRuntime.Runtime.ThreadPoolDispatcher (System.Func`1[TResult] callback) [0x00006] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/ObjCRuntime/Runtime.cs:293
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00009] in /Library/Frameworks/Xamarin.iOS.framework/Versions/12.6.0.25/src/Xamarin.iOS/mcs/class/referencesource/mscorlib/system/threading/threadpool.cs:1210
```

**Comment by juanlaube** [comment]
The happened for 3 users in 1.22 and only a few in 1.21. Mvx (the navigation service in particular) is in the stack traces, and has been replaced in 2.0, unless someone has an idea, I'd like to move this to 2.1 and see if this is still a problem after mvexit.

About the stack trace in handled exceptions, AppCenter shows the box in the UI but the SDK doesn't allow us to pass the stack trace, I checked this in our code and their docs.

**Comment by arturs105** [comment]
There _are_ stack traces, however they aren't showing up in the general crash overview. You have to go  to to "Reports" section and check individual crashes. But wait, you say, the stack traces aren't there as well?! That's because you can only see them in the raw json data by clicking on "RAW" at the right side of the page.

**Comment by amulware** [comment]
Closing this as I can't find any crashes in either 2.2.1 - if this still exists it's very rare and we can open a fresh issue so we can leave all the baggage here behind.
